refactor(router): Produce error message when canActivate is used with…

atscott opened the pull request:
… redirectTo

Redirects in the router are processed before activations. This means that a canActivate will
never execute if a route has a redirect. Rather than silently ignoring
the invalid config, developers should be notified so they know why it
doesn't work.

Closes #18605
The feature request for a function/class redirect is covered in #13373.

Note that this _is_ a new error that would prevent routing from working. We may want to discuss the need to land this along with a migration that removes `canActivate` from configs with `redirectTo` if this is found to be a common occurrence. Even without a migration, it may still be necessary to land this in a minor/major version to prevent new failures in patch version updates.

atscott committed:
refactor(router): Produce error message when canActivate is used with redirectTo

Redirects in the router are processed before activations. This means that a canActivate will
never execute if a route has a redirect. Rather than silently ignoring
the invalid config, developers should be notified so they know why it
doesn't work.

Closes #18605
The feature request for a function/class redirect is covered in #13373.

mhevery reviewed (changes requested):
Can we include a test?

atscott commented:
> Can we include a test?

Good call - done.

atscott commented:
[presubmit](https://test.corp.google.com/ui#id=OCL:346797394:BASE:346797424:1607620384087:48c13972)

atscott commented:
Initial presubmit show at least two problematic configs in the subset of tests run. 
One important note when making the decision on when/how to land this: This check is guarded by the `ngDevMode` flag so it would not cause failures in production for applications that have this flag correctly configured.

atscott committed:
feat(core): Add schematic to fix invalid `Route` configs

`Route` configs with `redirectTo` as well as `canActivate` are not valid
because the `canActivate` guards will never execute. Redirects are
applied before activation. There is no error currently for these
configs, but another commit will change this so that an error does
appear in dev mode. This migration fixes the configs by removing the
`canActivate` property.

atscott commented:
@mhevery - There were 5 places globally in g3 where this produced new errors. Again, these errors would only happen in dev mode. Still, I've updated the label to target a `minor` release.

I've also added a commit that includes a migration for this change. @crisbeto - would you mind reviewing the schematic change?

crisbeto commented on packages/core/schematics/migrations/can-activate-with-redirect-to/index.ts:
```diff
@@ -0,0 +1,56 @@
+  };
+}
+
+function runNavigationExtrasOmissionsMigration(tree: Tree, tsconfigPath: string, basePath: string) {
```
The function name seems incorrect since it's referring to the navigation extras migration.

atscott replied:
👍

crisbeto commented on packages/core/schematics/migrations/can-activate-with-redirect-to/util.ts:
```diff
@@ -0,0 +1,63 @@
+    if ((ts.isPropertyAssignment(property) || ts.isShorthandPropertyAssignment(property)) &&
+        (ts.isStringLiteralLike(property.name) || ts.isNumericLiteral(property.name) ||
+         ts.isIdentifier(property.name))) {
+      if (property.name.text === propertyName) {
```
Couldn't we merge this `if` statement into the one above, after it has asserted that we have the correct node type?

crisbeto commented on packages/router/src/utils/config.ts:
```diff
@@ -58,6 +58,12 @@ function validateNode(route: Route, fullPath: string): void {
       throw new Error(`Invalid configuration of route '${
           fullPath}': redirectTo and component cannot be used together`);
     }
+    if (route.redirectTo && route.canActivate) {
```
Should we guard this with `ngDevMode`?

atscott replied:
All of these checks are within a guarded block above.

crisbeto reviewed (approved):
LGTM on the migration code with a few nits.

atscott committed:
fixup! feat(core): Add schematic to fix invalid `Route` configs

atscott committed:
fixup! feat(core): Add schematic to fix invalid `Route` configs

mhevery reviewed (approved):
LGTM for router (not familiar with schematics)

atscott commented:
[presubmit](http://test/OCL:346797394:BASE:348035738:1608226255862:c6ad6c97)

josephperrott reviewed (approved):
LGTM

Reviewed-for: dev-infra
